File: web/src/api/ingest.ts
import api from './client'

// =====================================
// LazyIngest · 一键入库 API
// =====================================
//
// 该模块对接后端 /api/admin/ingest/* 系列接口：
//   - 用户只给 source_path（也可显式指定 target_root），系统自动完成
//     扫描 → AI 分类 → 命名 → 落盘 → 建库 → 扫描全流程；
//   - 默认 hardlink 优先，跨卷自动 copy；不删除源文件；
//   - 进度通过 GET /jobs/:id 轮询，或订阅 WS 事件 ingest_progress。

export type IngestJobStatus =
  | 'pending'
  | 'scanning'
  | 'classifying'
  | 'planning'
  | 'executing'
  | 'completed'
  | 'failed'
  | 'canceled'

export type NamingStyle = 'jellyfin' | 'plex'

export interface IngestStats {
  scanned: number
  classified: number
  planned: number
  executed: number
  skipped: number
  failed: number
  unsorted: number
  /** 视觉占用字节数（含 hardlink） */
  bytes_logical?: number
  /** 真实占用字节数（hardlink 为 0） */
  bytes_physical?: number
}

export interface IngestJob {
  id: string
  source_path: string
  target_root: string
  keep_original: boolean
  naming_style: NamingStyle | string
  /** 落盘模式 hardlink / move */
  mode?: 'hardlink' | 'move' | string
  /** Move 模式下的 journal 文件路径 */
  journal_path?: string
  status: IngestJobStatus
  phase: string
  progress: number
  /** JSON 字符串：本次自动创建/复用的媒体库 ID 数组 */
  library_ids: string
  /** JSON 字符串：本次产出的 RenamePlan ID 数组 */
  plan_ids: string
  /** JSON 字符串：IngestStats */
  stats: string
  error_message: string
  created_by: string
  created_at: string
  updated_at: string
  started_at: string | null
  completed_at: string | null
}

export interface ExistingIngestRef {
  job_id: string
  target_root: string
  mode: string
  completed_at: string
}

/** 409 重复入库响应体 */
export interface AlreadyIngestedError {
  error: string
  code: 'already_ingested'
  existing_jobs: ExistingIngestRef[]
}

export interface RollbackResult {
  total: number
  restored_mv: number
  skipped_mv: number
  removed_dir: number
  kept_dir: number
  errors: string[]
  corrupted: number
}

export interface SubmitRequest {
  source_path: string
  /** 可选：默认 = source_path/_organized */
  target_root?: string
  /** 可选：默认 jellyfin */
  naming_style?: NamingStyle
  /** 可选：落盘模式 hardlink（默认，零风险）/ move（专家模式，原地移动） */
  mode?: 'hardlink' | 'move'
  /** 可选：强制重新入库（跳过同源检测） */
  force?: boolean
}

export const ingestApi = {
  submit: (req: SubmitRequest) =>
    api.post<{ data: IngestJob }>('/admin/ingest/submit', req),

  listJobs: (limit = 50) =>
    api.get<{ data: IngestJob[] }>('/admin/ingest/jobs', { params: { limit } }),

  getJob: (id: string) =>
    api.get<{ data: IngestJob }>(`/admin/ingest/jobs/${id}`),

  cancelJob: (id: string) =>
    api.post<{ message: string }>(`/admin/ingest/jobs/${id}/cancel`),

  /** 取该 Job 关联的所有文件级明细（来自 RenamePlanItem） */
  getJobItems: (id: string) =>
    api.get<{ data: IngestJobItem[]; total: number }>(`/admin/ingest/jobs/${id}/items`),

  /** 回滚 Move 模式任务：倒序还原文件位置 */
  rollback: (id: string) =>
    api.post<{ data: RollbackResult }>(`/admin/ingest/jobs/${id}/rollback`),
}

/**
 * IngestJobItem 单条文件明细
 *
 * 字段对齐后端 model.RenamePlanItem，仅取 UI 需要的部分。
 */
export interface IngestJobItem {
  id: string
  plan_id: string
  source_path: string
  source_name: string
  target_path: string
  target_name: string
  parsed_title: string
  parsed_year: number
  media_type: 'movie' | 'episode' | 'unknown' | string
  season_num: number
  episode_num: number
  confidence: number
  ai_invoked: boolean
  safety_ok: boolean
  safety_note: string
  status: 'pending' | 'skipped' | 'unsafe' | 'executed' | 'failed' | 'reverted' | string
  error_msg: string
  excluded: boolean
  created_at: string
}

/** 文件状态中文映射 */
export function itemStatusLabel(s: string): string {
  switch (s) {
    case 'executed':
      return '已落盘'
    case 'skipped':
      return '已跳过'
    case 'unsafe':
      return '安全检测拦截'
    case 'failed':
      return '失败'
    case 'pending':
      return '待执行'
    case 'reverted':
      return '已回滚'
    default:
      return s
  }
}

/** 给前端按状态分组渲染 */
export function groupItemsByStatus(items: IngestJobItem[]): Record<string, IngestJobItem[]> {
  const out: Record<string, IngestJobItem[]> = {
    failed: [],
    unsafe: [],
    skipped: [],
    executed: [],
    pending: [],
  }
  for (const it of items) {
    if (!out[it.status]) out[it.status] = []
    out[it.status].push(it)
  }
  return out
}

// 解析 stats JSON 字符串为对象（容错处理）
export function parseIngestStats(job: IngestJob | null | undefined): IngestStats {
  const empty: IngestStats = {
    scanned: 0,
    classified: 0,
    planned: 0,
    executed: 0,
    skipped: 0,
    failed: 0,
    unsorted: 0,
    bytes_logical: 0,
    bytes_physical: 0,
  }
  if (!job?.stats) return empty
  try {
    const parsed = JSON.parse(job.stats) as Partial<IngestStats>
    return { ...empty, ...parsed }
  } catch {
    return empty
  }
}

/** 人类可读的字节数格式 */
export function formatBytes(n: number | undefined | null): string {
  if (!n || n <= 0) return '0 B'
  const units = ['B', 'KB', 'MB', 'GB', 'TB']
  let i = 0
  let v = n
  while (v >= 1024 && i < units.length - 1) {
    v /= 1024
    i++
  }
  return `${v.toFixed(v >= 100 ? 0 : 1)} ${units[i]}`
}

// 解析 library_ids JSON 字符串
export function parseLibraryIds(job: IngestJob | null | undefined): string[] {
  if (!job?.library_ids) return []
  try {
    const arr = JSON.parse(job.library_ids)
    return Array.isArray(arr) ? (arr as string[]) : []
  } catch {
    return []
  }
}

// 任务是否处于"运行中"状态
export function isJobRunning(job: IngestJob | null | undefined): boolean {
  if (!job) return false
  return ['pending', 'scanning', 'classifying', 'planning', 'executing'].includes(job.status)
}

// 中文化任务状态（仅用于展示）
export function statusLabel(status: IngestJobStatus): string {
  switch (status) {
    case 'pending':
      return '排队中'
    case 'scanning':
      return '扫描源目录'
    case 'classifying':
      return 'AI 分类'
    case 'planning':
      return '生成命名计划'
    case 'executing':
      return '整理文件'
    case 'completed':
      return '已完成'
    case 'failed':
      return '失败'
    case 'canceled':
      return '已取消'
    default:
      return status
  }
}

File: internal/model/ingest.go
package model

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// ==================== 懒人入库（Lazy Ingest）任务 ====================
//
// 该模型只为"一键入库"流程提供任务编排与进度跟踪：
//   1) 用户只给一个源目录 SourcePath；
//   2) 系统自动扫描 -> AI 分类 -> 建库 -> 生成命名规划 -> 落盘；
//   3) 全过程进度、统计、错误均沉淀在该任务记录里，供前端轮询/订阅。
//
// 安全约束：
//   - 所有产出的文件最终落到 SourcePath/_organized/ 之下（懒人模式 A 方案）；
//   - 默认 KeepOriginal=true（hardlink 优先；不可时降级 copy，绝不删除源）；
//   - 置信度低于阈值的条目挂在 RenamePlan 中等待人工，绝不自动落盘。

// IngestJobStatus 懒人入库任务状态
type IngestJobStatus string

const (
	IngestJobStatusPending     IngestJobStatus = "pending"     // 已创建，等待执行
	IngestJobStatusScanning    IngestJobStatus = "scanning"    // 扫描磁盘+落库
	IngestJobStatusClassifying IngestJobStatus = "classifying" // AI 分类
	IngestJobStatusPlanning    IngestJobStatus = "planning"    // 生成 RenamePlan
	IngestJobStatusExecuting   IngestJobStatus = "executing"   // 落盘移动
	IngestJobStatusCompleted   IngestJobStatus = "completed"   // 完成
	IngestJobStatusFailed      IngestJobStatus = "failed"      // 失败
	IngestJobStatusCanceled    IngestJobStatus = "canceled"    // 用户取消
)

// IngestJobMode 落盘策略
//   - hardlink: 默认，仅创建硬链接，源文件 0 风险，零额外空间（懒人模式）
//   - move    : 专家模式，os.Rename 原地移动；同卷瞬间完成、释放源；提供 journal 回滚
type IngestJobMode string

const (
	IngestJobModeHardlink IngestJobMode = "hardlink"
	IngestJobModeMove     IngestJobMode = "move"
)

// IngestJob 懒人入库任务
//
// 字段尽量小而精，只保留"一键入库"必要的编排字段；明细放在 RenamePlan / RenamePlanItem。
type IngestJob struct {
	ID string `json:"id" gorm:"primaryKey;type:text"`

	// 输入：用户唯一需要给的参数
	SourcePath string `json:"source_path" gorm:"type:text;not null"`

	// 自动派生：目标根目录（默认 = SourcePath/_organized）
	TargetRoot string `json:"target_root" gorm:"type:text"`

	// 执行策略（懒人模式默认值即可，全部硬编码到服务里，模型仅记录）
	KeepOriginal bool   `json:"keep_original" gorm:"default:true"` // 默认不删源
	NamingStyle  string `json:"naming_style" gorm:"type:text;default:jellyfin"`

	// 落盘模式：hardlink（默认，懒人）/ move（专家，原地移动 + journal 回滚）
	Mode string `json:"mode" gorm:"type:text;default:hardlink;index"`
	// Move 模式下的 journal 文件绝对路径（每行一条 JSON 操作记录，用于回滚）
	JournalPath string `json:"journal_path" gorm:"type:text"`

	// 状态机
	Status IngestJobStatus `json:"status" gorm:"type:text;default:pending;index"`
	Phase  string          `json:"phase" gorm:"type:text"` // 人类可读阶段（与 Status 对应，便于 UI 展示）

	// 进度（0-100，整体感知）
	Progress int `json:"progress"`

	// 派生产物
	LibraryIDs string `json:"library_ids" gorm:"type:text"` // JSON 数组：本次自动创建/复用的媒体库 ID
	PlanIDs    string `json:"plan_ids" gorm:"type:text"`    // JSON 数组：本次产出的 RenamePlan ID

	// 统计快照（JSON）
	// {"scanned":N,"classified":N,"planned":N,"executed":N,"skipped":N,"failed":N,"unsorted":N}
	Stats string `json:"stats" gorm:"type:text"`

	// 错误信息（最后一条致命错误）
	ErrorMessage string `json:"error_message" gorm:"type:text"`

	// 创建者（用户 ID，多用户审计）
	CreatedBy string `json:"created_by" gorm:"type:text"`

	CreatedAt   time.Time      `json:"created_at" gorm:"index"`
	UpdatedAt   time.Time      `json:"updated_at"`
	StartedAt   *time.Time     `json:"started_at"`
	CompletedAt *time.Time     `json:"completed_at"`
	DeletedAt   gorm.DeletedAt `json:"-" gorm:"index"`
}

// TableName 显式表名
func (IngestJob) TableName() string {
	return "ingest_jobs"
}

// BeforeCreate 自动生成 UUID
func (j *IngestJob) BeforeCreate(tx *gorm.DB) error {
	if j.ID == "" {
		j.ID = uuid.New().String()
	}
	return nil
}
